composer/utils/collect_env.py

```diff
@@ -0,0 +1,245 @@
+
+Or manually as a standalone script:
+
+.. code-block::
+
+    python composer/utils/collect_env.py
+
+To generate a system report from within a user application see :func:`print_env`.
+
+A custom excepthook wrapper is also provided which extends the default :func:`sys.excepthook`
+to automatically collect system information when an exception is raised.
+
+To override the default :func:`sys.excepthook` see :func:`configure_excepthook`.
+
+To restore the default :func:`sys.excepthook` see :func:`restore_excepthook`.
+"""
+
+import sys
+import time
+from typing import NamedTuple
+
+import cpuinfo
+import psutil
+
+__all__ = ["configure_excepthook", "print_env", "restore_excepthook"]
+
+# Check if PyTorch is installed
+try:
+    import torch.utils.collect_env as torchenv
+    from torch.cuda import get_device_name as accel_device_name
+    from torch.cuda import is_available as cuda_available
+    TORCH_AVAILABLE = True
+except (ImportError, NameError, AttributeError, OSError):
+    TORCH_AVAILABLE = False
+
+# Check if Composer is installed
+try:
+    import composer
+    from composer.utils import dist
+    COMPOSER_AVAILABLE = True
+except (ImportError, NameError, AttributeError, OSError):
+    COMPOSER_AVAILABLE = False
+
+COMPOSER_OPEN_ISSUE_URL = "https://github.com/mosaicml/composer/issues/new/choose"
+EXCEPTION_MSG = f"If you believe this exception was raised due to a Composer bug, " + \
+                f"file a bug report at: {COMPOSER_OPEN_ISSUE_URL}\n" + \
+                f"Please include the following environment report:\n"
+
+
+# Same convention as Torch collect_env, create a namedtuple to track collected fields
+class ComposerEnv(NamedTuple):
+    composer_version: str
+    number_of_nodes: int
+    host_processor_model_name: str
+    host_processor_core_count: int
+    accelerators_per_node: int
+    accelerator_model_name: str
+
+
+# Helper functions to get Composer environment information
+def get_composer_version() -> str:
+    """Query the Composer version."""
+    return str(composer.__version__)
+
+
+def get_num_nodes() -> int:
+    """Query the number of nodes."""
+    return int(dist.get_world_size() / dist.get_local_world_size())
+
+
+def get_host_processor_name() -> str:
+    """Query the host processor name."""
+    cpu_info = cpuinfo.get_cpu_info()
+    return str(cpu_info.get('brand_raw', "CPU"))
+
+
+def get_host_processor_cores() -> int:
+    """Determine the number of physical host processor cores."""
+    return psutil.cpu_count(logical=False)
+
+
+def get_accel_per_node() -> int:
+    """Determine the number of accelerators per node."""
+    return dist.get_local_world_size() if cuda_available() else 0
+
+
+def get_accel_model_name() -> str:
+    """Query the accelerator name."""
+    return accel_device_name(None) if cuda_available() else "N/A"
+
+
+# Excepthook wrapper, wraps default excepthook and prints env info
+def _excepthook_wrapper(type, value, tb) -> None:
+    sys.__excepthook__(type, value, tb)
```

Comment:
Instead of calling the fallback `__excepthook__`, could we capture the previous excepthook and call that? If the user or another package monkey-patched the excepthook, then we would not be calling that, but rather the default excepthook).